src/modules/Game.class.js

```diff
@@ -1,68 +1,263 @@
+
+    if (!emptyCells.length) {
+      return;
+    }
+
+    const [row, col] =
+      emptyCells[Math.floor(Math.random() * emptyCells.length)];
+
+    this.state[row][col] = Math.random() < 0.9 ? 2 : 4;
+  }
+
+  getEmptyCells() {
+    const matrix = this.getState();
+
+    return matrix
+      .flatMap((row, rowIndex) =>
+        // eslint-disable-next-line prettier/prettier
+        row.map((cel, colIndex) => (cel === 0 ? [rowIndex, colIndex] : null)))
+      .filter((cell) => cell !== null);
+  }
+
+  rotateLeft(transformedMatrix) {
+    const resultMatrix = [];
+    const cols = transformedMatrix[0].length;
+    const rows = transformedMatrix.length;
+
+    for (let col = 0; col < cols; col++) {
+      resultMatrix.push(Array.from({ length: rows }, () => ''));
+    }
+
+    for (let rowIndex = 0; rowIndex < rows; rowIndex++) {
+      for (let colIndex = 0; colIndex < cols; colIndex++) {
+        const newRow = rowIndex;
+        const newCol = cols - 1 - colIndex;
+
+        resultMatrix[newRow][newCol] = transformedMatrix[colIndex][rowIndex];
+      }
+    }
+
+    return resultMatrix;
+  }
+
+  rotateRight(matrix) {
+    const resultMatrix = [];
+    const cols = matrix[0].length;
+    const rows = matrix.length;
+
+    for (let col = 0; col < cols; col++) {
+      resultMatrix.push(Array.from({ length: rows }, () => ''));
+    }
+
+    for (let rowIndex = 0; rowIndex < rows; rowIndex++) {
+      for (let colIndex = 0; colIndex < cols; colIndex++) {
+        const newRow = cols - 1 - colIndex;
+        const newCol = rowIndex;
+
+        resultMatrix[newRow][newCol] = this.state[rowIndex][colIndex];
+      }
+    }
+
+    return resultMatrix;
+  }
+
+  isStateValid(currentState) {
+    if (this.status !== Game.gameStatuses.playing) {
+      return false;
+    }
+
+    for (const row of currentState) {
+      let hasAdjacentEqualCells = false;
+      let hasIsolatedEmptyCell = false;
+
+      for (let i = 0; i < row.length - 1; i++) {
+        if (row[i] === row[i + 1]) {
+          hasAdjacentEqualCells = true;
+          break;
+        }
+
+        if (!row[i]) {
+          hasIsolatedEmptyCell = true;
+        }
+      }
+
+      if (hasAdjacentEqualCells || hasIsolatedEmptyCell) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  completeMoveTasks(count = 1) {
+    for (let i = 0; i < count; i++) {
+      this.generateNewTile();
+    }
+
+    const state = this.getState();
+
+    if (this.isVictory(state)) {
+      this.status = Game.gameStatuses.win;
+    } else if (this.isDefeat(state)) {
+      this.status = Game.gameStatuses.lose;
+    }
+  }
+
+  isDefeat(state) {
+    const rotatedRight = this.rotateRight(state);
+
+    return [state, rotatedRight].every(
+      (currentState) => !this.isStateValid(currentState),
+    );
+  }
+
+  isVictory(state) {
+    return state.flat().some((tile) => tile === 2048);
+  }
+
+  resetState() {
+    this.state = [...this.initialState.map((row) => [...row])];
```

Comment:
The `resetState` method directly uses `this.initialState`, which could lead to unintended side effects if `initialState` is modified elsewhere. It would be safer to deep copy the `initialState` when resetting the state.